sympy/matrices/matrices.py

```diff
@@ -1581,6 +1583,12 @@ def _matrix_pow_by_recursion(self, num):
     def __pow__(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        num = sympify(num)
+        if not num.is_integer and (num >= 0) == True and not num % 1 == 0:
```

Comment:
We just need to allow **non-negative integers** and **floats which are integers** and the rest if (```not num.is_integer and (num >= 0) == True and not num % 1 == 0```) the computed jordan block is nilpotent of size > 1 we raise an error. Here ```self.eigenvals()[0]``` computes the size of the block with eigenvalue 0. These are the conditions and logic applied, please correct me if I'm wrong anywhere or I'm missing something.